cmd/lvldb2elastic/databasereader/databaseReader.go

```diff
@@ -0,0 +1,251 @@
+	if len(dbFilePath) == 0 {
+		return nil, ErrEmptyDbFilePath
+	}
+	if check.IfNil(marshalizer) {
+		return nil, ErrNilMarshalizer
+	}
+
+	return &databaseReader{
+		directoryReader: factory.NewDirectoryReader(),
+		marshalizer:     marshalizer,
+		dbFilePath:      dbFilePath,
+	}, nil
+}
+
+func (dr *databaseReader) GetDatabaseInfos() ([]*DatabaseInfo, error) {
+	dbs := make([]*DatabaseInfo, 0)
+	epochsDirectories, err := dr.directoryReader.ListDirectoriesAsString(dr.dbFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	re := regexp.MustCompile("[0-9]+")
+
+	for _, dirname := range epochsDirectories {
+		epochStr := re.FindString(dirname)
+		epoch, err := strconv.ParseInt(epochStr, 10, 64)
+		if err != nil {
+			log.Warn("cannot parse epoch number from directory name", "directory name", dirname)
+			continue
+		}
+
+		shardDirectories, err := dr.getShardDirectoriesForEpoch(filepath.Join(dr.dbFilePath, dirname))
+		if err != nil {
+			log.Warn("cannot parse shard directories for epoch", "epoch directory name", dirname)
+			continue
+		}
+
+		for _, shardDir := range shardDirectories {
+			dbs = append(dbs,
+				&DatabaseInfo{
+					Epoch: uint32(epoch),
+					Shard: shardDir,
+				})
+		}
+	}
+
+	if len(dbs) == 0 {
+		return nil, errors.New("no database found")
+	}
+
+	return dbs, nil
+}
+
+func (dr *databaseReader) getShardDirectoriesForEpoch(dirEpoch string) ([]uint32, error) {
+	shardIDs := make([]uint32, 0)
+	directoriesNames, err := dr.directoryReader.ListDirectoriesAsString(dirEpoch)
+	if err != nil {
+		return nil, err
+	}
+
+	shardDirs := make([]string, 0, len(directoriesNames))
+	for _, dirName := range directoriesNames {
+		isShardDir := strings.HasPrefix(dirName, "Shard_")
+		if !isShardDir {
+			continue
+		}
+
+		shardDirs = append(shardDirs, dirName)
+	}
+
+	for _, fileName := range shardDirs {
+		stringToSplitBy := "Shard_"
+		splitSlice := strings.Split(fileName, stringToSplitBy)
+		if len(splitSlice) < 2 {
+			continue
+		}
+
+		shardID := core.MetachainShardId
+		shardIdStr := splitSlice[1]
+		if shardIdStr != "metachain" {
+			u64, err := strconv.ParseUint(shardIdStr, 10, 32)
+			if err != nil {
+				continue
+			}
+			shardID = uint32(u64)
+		}
+
+		shardIDs = append(shardIDs, shardID)
+	}
+
+	return shardIDs, nil
+}
+
+// TODO : finish
+
+func (dr *databaseReader) GetHeaders(dbInfo *DatabaseInfo) ([]data.HeaderHandler, error) {
+	hdrStorer := shardBlocksStorer
+	if dbInfo.Shard == core.MetachainShardId {
+		hdrStorer = metaHeadersStorer
+	}
+
+	shardIDStr := fmt.Sprintf("%d", dbInfo.Shard)
+	if shardIDStr == fmt.Sprintf("%d", core.MetachainShardId) {
+		shardIDStr = "metachain"
+	}
+	persisterPath := filepath.Join(dr.dbFilePath, fmt.Sprintf("Epoch_%d", dbInfo.Epoch), fmt.Sprintf("Shard_%s", shardIDStr), hdrStorer)
+	options := &opt.Options{
+		// disable internal cache
+		BlockCacheCapacity:     -1,
+		OpenFilesCacheCapacity: maxOpenFiles,
+	}
+	lvlDb, err := leveldb2.OpenFile(persisterPath, options)
```

Comment:
I think this is a wrong approach. Why not using storer interface and implementations?